proposals/text/3959-json-converter.md

```diff
@@ -0,0 +1,121 @@
+
+For many such processing, json is a convenient intermediate format as it allows for things like specifying the metadata.
+
+Specifically, one use-case that has come up for a team using haystack: they want to use a PDF parser (for tables) currently not in haystack. As such, they want to handle the parsing themselves outside of haystack, put the parsed result into a json file, and then pass it onto a haystack API endpoint.
+
+Having a `JsonConverter` node would allow users to setup a haystack pipeline to ingest such data without the user having to create a custom node for it.
+
+# Detailed design
+
+The converter would primarily be a wrapper around `Document.from_dict`.
+
+```python
+class JsonConverter(BaseConverter):
+    def __init__(self, ...):
+        ...
+
+    def convert(
+        self,
+        file_path: Path,
+        meta: Optional[Dict[str, str]] = None,
+        encoding: Optional[str] = "UTF-8",
+        id_hash_keys: Optional[List[str]] = None,
+        ...
+    ) -> List[Document]:
+        if id_hash_keys is None:
+            id_hash_keys = self.id_hash_keys
+
+        documents = []
+        with open(file_path, encoding=encoding, errors="ignore") as f:
+            data = json.load(f)
+            for doc_dict in data:
+                doc_dict = dict(doc_dict)
+                doc_dict['id_hash_keys'] = id_hash_keys
+                doc_dict['meta'] = doc_dict.get('meta', dict())
+
+                if meta:
+                    doc_dict['meta'].update(meta)
+
+                documents.append(Document.from_dict(doc_dict))
+
+        return documents
+```
+
+# Drawbacks
+
+- It would add another node that needs to be maintained and documented.
+
+# Alternatives
+
+- This node could be created as a custom node for the particular application where it is required. But could be better to have it out-of-the-box.
+- Design Alternative: Also, provide options to map custom fields to `Document` fields (E.g. {"review": "content"}) which could make this node a bit more flexible and might mean the user doesn't have to do some pre-formatting beforehand. But this can be a future development.
+
+# Adoption strategy
+
+It doesn't introduce a breaking change and wouldn't require changes in existing pipelines.
+
+# How we teach this
+
+It would be good to have this be part of the Guide (perhaps under File Converters).
+
+Could also be mentioned in one of the tutorials. For instance, in the preprocessing tutorial where we say "Haystack expects data to be provided as a list documents in the following dictionary format".
+
+# Unresolved questions
+
+- Should the `content_type` allowed be restricted (e.g. only "text" and "table"). And relatedly, should the name be more specific? E.g. `JsonTableTextConverter` rather than `JsonConverter`. Currently leaning towards no restriction and the `JsonConverter` name.
```

Comment:
I agree that leaving the name as `JsonConverter` would be a good idea, especially with the idea of expanding this node to allow for custom field mapping in the future (as you mention under Alternatives).

Reply:
If we don't restrict the `content_type`, we should document how to pass (for example) audio data in the field `content_audio` for a document of type `audio` (base64?).

Reply:
I'd think it would mirror `Document`. My understanding is that for content_type `image` and `audio`, the `Document.content` field just contains the file_path(?). So the json input here would also do the same.

(Though I do see how having a self-sufficient file would be nice).

Reply:
I agree we don't need to limit the content_types. right now "file" documents (image, audio) are quite shallow and only holds  a path, so a user could totally use this node to load a bunch of image Documents if they wish so. No need to introduce an artificial limitation.